erpnext/accounts/doctype/payment_request/payment_request.py

```diff
@@ -0,0 +1,190 @@
+		
+		return jv
+		
+	def send_email(self):
+		"""send email with payment link"""
+		frappe.sendmail(recipients=self.email_to, sender=None, subject=self.subject,
+			message=self.get_message(), attachments=[frappe.attach_print(self.reference_doctype, 
+			self.reference_name, file_name=self.reference_name, print_format=self.print_format)])
+						
+	def get_message(self):
+		"""return message with payment gateway link"""
+		return  cstr(self.message) + """ <a href="%s"> Click here to pay </a>"""%self.payment_url
+		
+	def set_failed(self):
+		pass
+	
+	def set_cancelled(self):
+		frappe.db.set_value(self.doctype, self.name, "status", "Cancelled")
+	
+	def make_communication_entry(self):
+		"""Make communication entry"""
+		comm = frappe.get_doc({
+			"doctype":"Communication",
+			"subject": self.subject,
+			"content": self.get_message(),
+			"sent_or_received": "Sent",
+			"reference_doctype": self.reference_doctype,
+			"reference_name": self.reference_name
+		})
+		comm.insert(ignore_permissions=True)
+
+@frappe.whitelist()
+def make_payment_request(**args):
+	"""Make payment request"""
+	args = frappe._dict(args)
+	ref_doc = get_reference_doc_details(args.dt, args.dn)
+	name, gateway, payment_account, message = get_gateway_details(args)
+	
+	pr = frappe.new_doc("Payment Request")
+	pr.update({
+		"payment_gateway": name,
+		"gateway": gateway,
+		"payment_account": payment_account,
+		"currency": ref_doc.currency,
+		"amount": get_amount(ref_doc, args.dt),
+		"mute_email": args.mute_email or 0,
+		"email_to": args.recipient_id or "",
+		"subject": "Payment Request for %s"%args.dn,
+		"message": message,
+		"reference_doctype": args.dt,
+		"reference_name": args.dn
+	})
+	
+	if args.return_doc:
+		return pr
+		
+	if args.submit_doc:
+		pr.insert(ignore_permissions=True)
+		pr.submit()
+		return pr
+	
+	return pr.as_dict()
+
+def get_reference_doc_details(dt, dn):
+	""" return reference doc Sales Order/Sales Invoice"""
+	return frappe.get_doc(dt, dn)
+
+def get_amount(ref_doc, dt):
+	"""get amount based on doctype"""
+	if dt == "Sales Order":
+		amount = flt(ref_doc.base_grand_total) - flt(ref_doc.advance_paid)
+
+	if dt == "Sales Invoice":
+		amount = abs(ref_doc.outstanding_amount)
+	
+	if amount > 0:
+		return amount
+	else:
+		frappe.throw(_("Payment Entry is already created"))
+		
+def get_gateway_details(args):
+	"""return gateway and payment account of default payment gateway"""
+	if args.payemnt_gateway:
+		return frappe.db.get_value("Payment Gateway Account", args.payemnt_gateway, 
+			["name", "gateway", "payment_account", "message"])
+		
+	return frappe.db.get_value("Payment Gateway Account", {"is_default": 1}, 
+		["name", "gateway", "payment_account", "message"])
+
+@frappe.whitelist()
+def get_print_format_list(ref_doctype):
+	print_format_list = ["Standard"]
+	
+	print_format_list.extend(list(chain.from_iterable(frappe.db.sql("""select name from `tabPrint Format` 
+		where doc_type=%s""", ref_doctype, as_list=1))))
+	
```

Comment:
`[p.name for p in frappe.get_all("Print Format", filters={"doc_type": ref_doctype})]`